lib/spack/spack/build_systems/rocm.py

```diff
@@ -90,9 +90,10 @@ class ROCmPackage(PackageBase):
     # https://llvm.org/docs/AMDGPUUsage.html
     # Possible architectures
     amdgpu_targets = (
-        'gfx701', 'gfx801', 'gfx802', 'gfx803',
-        'gfx900', 'gfx906', 'gfx908', 'gfx90a', 'gfx1010',
-        'gfx1011', 'gfx1012'
+        'gfx701', 'gfx801', 'gfx802', 'gfx803', 'gfx900',
+        'gfx906', 'gfx908', 'gfx90a',
+        'gfx906:xnack-', 'gfx908:xnack-', 'gfx90a:xnack-', 'gfx90a:xnack+',
+        'gfx1010', 'gfx1011', 'gfx1012', 'gfx1030',
     )
```

Comment:
> A couple notable amdgpu_targets you're missing are:
> 
>     * `gfx900:xnack-` which is used by pretty much all the ROCm math libraries except rocBLAS and rocSOLVER
> 
>     * `gfx1031` which is not officially supported, but will probably work for some libraries

Thanks for pointing that out! After your comment I grepped spack to search for currently used architectures. Here's the output:

```
➜  spack git:(alby/rocm_amdgpu_targets) ✗ git grep -ho "gfx[0-9]\+[^ ,;'\`\".]*" | sort | uniq

gfx000
gfx1010
gfx1011
gfx1012
gfx1030
gfx1031
gfx700
gfx701
gfx801
gfx802
gfx803
gfx900
gfx900:xnack-
gfx902
gfx906
gfx906:xnack-
gfx908
gfx908:xnack-
gfx90a
gfx90a:xnack+
gfx90a:xnack-
```

It caught my eye `gfx000`, which is mentioned just in `rocm-tensile` package, but I cannot find it in the official list you mentioned. I'm not able to say if it is a typo or not.

About `TargetID` class you are proposing, it can be added in the same file where `ROCmPackage` is defined.
However, I would not be against having them explicitly listed. I don't know exactly how much complicated the logic for composing architectures can be. For instance, AFAICT, not all features makes sense for all targets, so a filtering may be needed.

Reply:
`gfx000` is used as a null value by rocminfo. It has been repurposed to have various meanings in different contexts.

> However, I would not be against having them explicitly listed. I don't know exactly how much complicated the logic for composing architectures can be. For instance, AFAICT, not all features makes sense for all targets, so a filtering may be needed.

My reasoning to use a class would be to model that logic (and other related properties of the GPU architecture). However, it seems the total number of possible target-id values is lower than I expected, as [only sramecc and xnack can be specified via target-id](https://github.com/RadeonOpenCompute/llvm-project/blob/rocm-5.1.1/clang/lib/Basic/TargetID.cpp#L29-L32). I've generated all valid values of target-id for gfx7, gfx8, gfx9, gfx10 and gfx11 based on the [clang AMD CGN GPUInfo table](https://github.com/RadeonOpenCompute/llvm-project/blob/1badba1e749eb8b21640d3d7e8addbd7b1128838/llvm/lib/Support/TargetParser.cpp#L64-L123):
```
gfx700
gfx701
gfx702
gfx703
gfx704
gfx705
gfx801
gfx801:xnack+
gfx801:xnack-
gfx802
gfx803
gfx805
gfx810
gfx810:xnack+
gfx810:xnack-
gfx900
gfx900:xnack+
gfx900:xnack-
gfx902
gfx902:xnack+
gfx902:xnack-
gfx904
gfx904:xnack+
gfx904:xnack-
gfx906
gfx906:xnack+
gfx906:xnack-
gfx906:sramecc+
gfx906:sramecc+:xnack+
gfx906:sramecc+:xnack-
gfx906:sramecc-
gfx906:sramecc-:xnack+
gfx906:sramecc-:xnack-
gfx908
gfx908:xnack+
gfx908:xnack-
gfx908:sramecc+
gfx908:sramecc+:xnack+
gfx908:sramecc+:xnack-
gfx908:sramecc-
gfx908:sramecc-:xnack+
gfx908:sramecc-:xnack-
gfx909
gfx909:xnack+
gfx909:xnack-
gfx90a
gfx90a:xnack+
gfx90a:xnack-
gfx90a:sramecc+
gfx90a:sramecc+:xnack+
gfx90a:sramecc+:xnack-
gfx90a:sramecc-
gfx90a:sramecc-:xnack+
gfx90a:sramecc-:xnack-
gfx90c
gfx90c:xnack+
gfx90c:xnack-
gfx940
gfx940:xnack+
gfx940:xnack-
gfx940:sramecc+
gfx940:sramecc+:xnack+
gfx940:sramecc+:xnack-
gfx940:sramecc-
gfx940:sramecc-:xnack+
gfx940:sramecc-:xnack-
gfx1010
gfx1010:xnack+
gfx1010:xnack-
gfx1011
gfx1011:xnack+
gfx1011:xnack-
gfx1012
gfx1012:xnack+
gfx1012:xnack-
gfx1013
gfx1013:xnack+
gfx1013:xnack-
gfx1030
gfx1031
gfx1032
gfx1033
gfx1034
gfx1035
gfx1036
gfx1100
gfx1101
gfx1102
gfx1103
```

Reply:
@cgmb can https://github.com/spack/spack/pull/30582/commits/023c5aa2954e533f6a1680c7c7419fc19e55211a be a reasonable starting point?

Reply:
Wow! It's like you read my mind. I had to do a double-take, because that looks so similar to what I was prototyping myself. I'll give it a more thorough review in a bit, but that looks great.

Reply:
😄 nice to hear that!

In the meanwhile I added another change...just a proposal as the rest, but it can be helpful: given a target id with a minimal set of features, returns the list of all canonical target ids satisfying it.

e.g. `gfx940:xnack+` -> `['gfx940:sramecc-:xnack+', 'gfx940:sramecc+:xnack+']`

It can also be extended to address the case of the empty input string as a shortcut for listing all available combination of archs/features.

Looking forward to your comments and indications on how to carry on this PR. 😉